src/main/java/ev3dev/utils/SysfsNIO.java

```diff
@@ -0,0 +1,250 @@
+package ev3dev.utils;
+
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.*;
+
+/**
+ * The class responsible to interact with SysfsNIO on EV3Dev
+ *
+ * @author Juan Antonio Breña Moral
+ */
+public class SysfsNIO {
+    private static final Logger log = org.slf4j.LoggerFactory.getLogger(SysfsNIO.class);
+    private static final ByteBuffer buffer = ByteBuffer.allocate(256);
```

Comment:
Maybe, it could generate some concurrency issues?

Reply:
All public methods are `synchronized` now, so this shouldn't be an issue. A stateless variant is also possible. However, that would likely mean allocating buffers on each read.

Reply:
Instead of synchronized access to that byte buffer it'd probably be better to promote responsibility for concurrency out of the Sysfs API. It'd be very useful for concurrent code to try to read several files at once. Concurrent hard looping access was my first hack around the performance problem, and it worked reasonably well.

Instead of allocating a new ByteBuffer for each read - allocate one per file - preferably in an object that persists and is owned by something not Sys* . That avoids the possible collision here, and will let you put the usual Java caveats about concurrency in the class responsible for accessing the device.

Reply:
> It'd be very useful for concurrent code to try to read several files at once. Concurrent hard looping access was my first hack around the performance problem, and it worked reasonably well.

My intuition here was based on the fact that EV3 has only one CPU core. From this viewpoint, there shouldn't be much benefit in reading the files in parallel (as reads/writes in sysfs should basically be CPU-bound kernel function calls). However, the `synchronized` here does serialize all accesses and parsing in sysfs and if multiple parallel reads helped previously, then that approach could be hindered by this change.

> Instead of synchronized access to that byte buffer it'd probably be better to promote responsibility for concurrency out of the Sysfs API.

This can be done only in some places. For example, reading the tacho counter from a motor is a simple `getTachoCount()` call without any arguments. The whole motor object could be specified to be usable only from one thread, but that does introduce new assumptions/restrictions to the API (although that could be OK). Another option is ThreadLocal<>, but I've removed it for the same reason - it assumes that there will not be a large number of distinct long-lived threads accessing Sysfs. Looking back though, it does not look very likely that there will be that many long-lived threads. I also found that objects *are* freed from ThreadLocal<> if their owning thread dies, just not if that thread is still alive because it is part of a thread pool.

I'm starting to think that the additional allocations could be OK after all. With multiple allocation sizes, the impact might not be that drastic, but it does penalize larger reads by doing the read twice.